texk/texlive/getnonfreefonts.man
.TH getnonfreefonts 1 "January 2007" "TeXLive" "TeXLive"
.SH NAME
getnonfreefonts \- download and install fonts which cannot be
distributed on a CD
.LP
getnonfreefonts-sys \- download and install fonts system\-wide which
cannot be distributed on a CD

.SH SYNOPSIS

.BI "getnonfreefonts [" options "] [" fontlist ]
.sp
.BI "getnonfreefonts-sys [" options "] [" fontlist ]

.SH DESCRIPTION
Some font suppliers provided fonts which can be used freely but they
disallow that money is charged for the distribution.  Hence, for
distributions on CD or DVD it is usually impossible to include
these fonts.

.B "getnonfreefonts"
is a script which downloads and installs these fonts.  Actually, it
first downloads the program
.BI "getfont"<year>
from tug.org which is then executed.
.BI "getfont"<year>
does the real work.  It will be updated occasionally to support more
fonts.  
.I <year>
denotes the TeXLive release.

.B getnonfreefonts
and
.B getnonfreefonts-sys
are identical except that
.B getnonfreefonts
installs files to $TEXMFHOME
while
.B getnonfreefonts-sys
installs files to $TEXMFLOCAL.

.BI "getfont"<year>
will not be installed permanently on your system.  It only
exists in a temporary directory while 
.BI getnonfreefonts
is running.

Run
.B getnonfreefonts --lsfonts
to get a list of supported fonts.

.SH OPTIONS

.TP
.BR "-a" , " --all"
.rg
Install all fonts.

.TP
.BR "-d" , " --debug"
.rg
Provide additional messages for debugging.  These messages are sent to
STDERR.

.TP
.BR "-f" , " --force"
.rg
Install fonts even if they are installed already.

.TP
.BR "-h" , " --help"
.rg
Print a help message.  Nothing is installed but the directories where
the fonts will be installed and the temporary directory are shown.
.BI "<" PID ">" 
will be replaced by the process ID of the current
process. 

.TP
.BR "-l" , " --lsfonts"
.rg
List all available fonts and tell whether they are installed already. 

.TP
.B --sys
.rg
If you call
.B getnonfreefonts
with the
.B --sys
option it behaves exactly like
.BR "getnonfreefonts-sys".

.TP
.BR "-v" , " --verbose"
.rg
Be more verbose.  This option is only needed for debugging.

.TP
.BR "--version"
.rg
Print version number.

.SH ENVIRONMENT
One of the environment variables
.BR "TMP" , " TEMP" ,
or 
.B  TMPDIR 
must exist.  They are used to determine where to put the temporary
files.  The directory must exist and must be writable.
.B /tmp
will be used if none of the variables is defined.

.SH REQUIREMENTS
Though this program is supposed to be used with 
.B TeX Live
it might work with other TeX distributions too.  The programs
.BR "wget"
or 
.BR "curl" ,
.BR "unzip" , 
and
.B kpsewhich
have to be installed and a recent version of
.B updmap 
is required which supports the 
.B --enable
option.

.SH BUGS
Please send bug reports to Reinhard Kotucha <reinhard.kotucha@web.de>.


ChangeLog
2008-06-09  Karl Berry  <karl@tug.org>

	* Build: try using find instead of ls, to avoid short param list
	on hp (Tim tlbuild mail, 6 Jun 2008 10:24:41).

2008-06-08  Karl Berry  <karl@tug.org>

	* utils/README: warn about nohup specifically.

2008-05-29  Karl Berry  <karl@tug.org>

	* Makefile.in (install): do not call selfautofix, we now keep
	kpathsea/texmf.cnf as a slave copy of Master/texmf/web2c.
	* selfautofix: remove file.

2008-05-28  Karl Berry  <karl@tug.org>

	* Makefile.in (install): invoke new run-texlinks target
	in texk/tetex.

	* Makefile.in (install-extra): do not go to omfonts, the omega.mk
	targets take care of that.
	Do not try to install release-tetex-src.txt.
	Do not try to run fmtutil, updmap, mktexlsr, texlinks or anything else.
	(world): since strip already depends on install, omit install.
	Add echo's for when targets are finished.	

2008-04-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.in: respect library dependencies.

2008-03-25  Karl Berry  <karl@tug.org>

	* withenable.ac: enable xindy; rebuild configure.
	* README,
	* utils/README: document clisp & libffcall prerequisites.

2008-03-24  Karl Berry  <karl@tug.org>

	* Build: exit with appropriate status

2008-03-09  Karl Berry  <karl@tug.org>

	* Build: exit if no binaries.

2008-03-08  Karl Berry  <karl@tug.org>

	* Build: --debug

2008-03-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* reautoconf: run first autoheader, then autoconf.

2008-02-29  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.in: update (AC_SET_MAKE -> AC_PROG_MAKE_SET).
	* tetex.ac: append --enable-tetex-build to args in addition
	to exporting enable_tetex_build=yes (for maintainer mode).
	* reautoconf: add autoheader, allow subdir selection.

2008-02-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* reautoconf: major rewrite.

2008-02-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tetex.ac: --enable-muliplatform modifies libdir.

	* withenable.ac (kpse_utils_pkglist): m4_defined list of all
	optional packages, used to generate --without-PKG options.
	* configure.in: use kpse_utils_pkglist to select subdirs.

2008-02-02  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Convert from autoconf-2.13 to autoconf-2.59+:
	* configure.in: Use KPSE_CONFIG_SUBDIRS.
	* Makefile.in: Add @configure_input@ line.
	* aclocal.m4 (new): Generated (aclocal -I texk/m4).
	* configure: Regenerated (autoconf).

2007-10-06  Karl Berry  <karl@tug.org>

	* Build (TL_WORKDIR): allow override of "Work".

2007-10-02  Karl Berry  <karl@tug.org>

	* configure.in: remove sam2p special case, since we omitted the
	program.

2007-08-14  Joachim Schrod  <jschrod@acm.org>

	* configure.in: dvipdfm and dvipdfmx need pnglib and zlib, too.

2007-07-16  Karl Berry  <karl@tug.org>

	* Makefile.in (lib): new target, depends on @LIBSDEP@.
	(all): do not depend on @LIBSDEP@, redundant since we already
		descend and build.
	From Joachim, 28 Jun 2007 11:47:10.

2007-03-21  Karl Berry  <karl@tug.org>

	* Makefile.in (distclean): remove mention of klibtool.

2006-12-14  Karl Berry  <karl@tug.org>

	* Build: allow invoking from anywhere, and more overrides.

2006-12-11  Karl Berry  <karl@tug.org>

	* Makefile.in (world, install): hmm, maybe better as a dependency
	than a submake.

	* Makefile.in (install): reinstate strip.
	(strip): actually strip.

2006-12-08  Karl Berry  <karl@tug.org>

	* Build: set GNUMAKE to TL_MAKE for freetype2/configure, and
	export it.  Reported by Manfred and Vladimir.

2006-12-05  Karl Berry  <karl@tug.org>

	* Build: no tlbuild.log, since then the exit status is always
	successful.

2006-11-08  Karl Berry  <karl@tug.org>

	* configure.in (PKGS): remove sam2p after all.

2006-07-12  Karl Berry  <karl@tug.org>

	* utils/sam2p/Makefile (install): use cp -p, not -a.

2006-06-02  Karl Berry  <karl@tug.org>

	* texk/web2c/omegadir/com16bit.ch, alephdir/com16bit-rc2.ch
	(sup_main_memory, sup_save_size): increase per tex.ch.
	From Akira.


texk/texlive/linked_scripts/getnonfreefonts.pl
#!/usr/bin/env perl

## getnonfreefonts
## Copyright 2006-2008 Reinhard Kotucha <reinhard.kotucha@web.de>
#
# This work may be distributed and/or modified under the
# conditions of the LaTeX Project Public License, either version 1.3
# of this license or (at your option) any later version.
# The latest version of this license is in
#   http://www.latex-project.org/lppl.txt
# 
# The current maintainer is Reinhard Kotucha.

my $TL_version='2008';

my $getfont_url="ftp://tug.org/tex/getnonfreefonts/getfont$TL_version";

use File::Copy;
use Getopt::Long;
$Getopt::Long::autoabbrev=0;

$opt_lsfonts=0;
$opt_force=0;

sub usage {
    print <<'EOF';
Usage:
    getnonfreefonts[-sys] [-a|--all] [-d|--debug] [-f|--force]
        [-l|--lsfonts] [-v|--verbose] [--version] [font1] [font2] ...

    getnonfreefonts installs fonts in $TEXMFHOME.
    getnonfreefonts-sys installs fonts in $TEXMFLOCAL.

    Options:
        -a|--all      Install all fonts.

        -d|--debug    Provide additional messages for debugging.

        -f|--force    Install fonts even if they are installed already.

        -h|--help     Print this message.

        -l|--lsfonts  List all fonts available.

        -v|--verbose  Be more verbose.

        --version     Print version number.

EOF
;
}

@ARGS=@ARGV;

GetOptions 
    "all|a",
    "debug|d",
    "force|f",
    "help|h",
    "lsfonts|l",
    "verbose|v",
    "version",
    "sys";

$^W=1 if $opt_debug;

my $sys=($0=~/-sys$/)? 1:0;
$sys=1 if $opt_sys;
$sys=1 if (defined $ENV{'TEX_SYS_PROG'});

@allpackages=@ARGV;

print "$TL_version\n" and exit 0 if $opt_version; 

if ($opt_help or !@ARGS) {
    print "\nThis is getnonfreefonts";
    print '-sys' if ($sys);
    print ", version $TL_version.\n\n";
    usage; 
}


sub message {
    my $message=shift;
    if ($message=~/^\[/) {
	print "$message\n";
    } else {
	printf "%-60s", $message;
    }
}


sub debug_msg {
    my $message=shift;
    if ($opt_debug) {
	print STDERR "DEBUG: $message.\n";
    }
}

sub which {
    my $prog=shift;
    my @PATH;
    my $PATH=$ENV{'PATH'};
    @PATH=split ':', $PATH;
    for my $dir (@PATH) {
	if (-x "$dir/$prog") {
	    return "$dir/$prog";
	}
    }
    return 0;
}


sub expand_var {
    my $var=shift;

    if ($^O=~/^MSWin(32|64)/i) {
	open KPSEWHICH, 'kpsewhich --expand-var=$'  . "$var |";
    } else {
	open KPSEWHICH, 'kpsewhich --expand-var=\$' . "$var |";
    }
    while (<KPSEWHICH>) {
	chop;
	return "$_";
    }
    close KPSEWHICH;
}

# get TMP|TEMP|TMPDIR environment variable

if ($opt_debug) {
    for my $var ("TMP","TEMP","TMPDIR") {
	if (defined $ENV{$var}) {
	    debug_msg "Environment variable $var='$ENV{$var}'";
	} else {
	    debug_msg "Environment variable $var not set";
	}
    }
}


if (defined $ENV{'TMP'}) {
    $SYSTMP="$ENV{'TMP'}";
} elsif (defined $ENV{'TEMP'}) {
    $SYSTMP="$ENV{'TEMP'}";
} elsif (defined $ENV{'TMPDIR'}) {
    $SYSTMP="$ENV{'TMPDIR'}";
} else {
    $SYSTMP="/tmp";
}

debug_msg "Internal variable SYSTMP set to '$SYSTMP'";

# Windows usually uses backslashes though forward slashes are probably
# allowed.  Perl always needs forward slashes.

# We convert backslashes to forward slashes on Windows.
$SYSTMP=~s/\\/\//g if ($^O=~/^MSWin(32|64)/i);


debug_msg "Internal variable SYSTMP converted to '$SYSTMP'";

sub check_tmpdir{
die "! ERROR: The temporary directory '$SYSTMP' doesn't exist.\n"
    unless (-d "$SYSTMP");

die "! ERROR: The temporary directory '$SYSTMP' is not writable.\n"
    unless (-w "$SYSTMP");
}

# Determine INSTALLROOT.

if ($opt_debug) {
    my @path=split ':', $ENV{PATH};
    for my $dir (@path) {
	debug_msg "Search for kpsewhich in '$dir'";
	opendir DIR, "$dir";
	my @kpsewichs=grep /^kpsewhich/, readdir DIR;
	closedir DIR;
	for (@kpsewichs) {
	    debug_msg "Found $dir/$_" if /^kpsewhich/;
	}
    }
}


$INSTALLROOTNAME=($sys)? 'TEXMFLOCAL':'TEXMFHOME';

$INSTALLROOT=expand_var "$INSTALLROOTNAME";

($sys)? debug_msg "sys=true":debug_msg "sys=false";

debug_msg "INSTALLROOT='$INSTALLROOT'";

$INSTALLROOT=~s/\\/\//g if ($^O=~/^MSWin(32|64)/i);

debug_msg "Internal variable INSTALLROOT converted to '$INSTALLROOT'";

sub check_installroot {
    mkdir "$INSTALLROOT" unless (-d "$INSTALLROOT");

    die "! ERROR: The variable $INSTALLROOTNAME is not set.\n"
	unless length ("$INSTALLROOT") > 0;

    die "! ERROR: The install directory '$INSTALLROOT' doesn't " .
	"exist.\n" .
	"         If this is the correct path, please create " .
	"this directory manually.\n" 
	unless (-d "$INSTALLROOT");
    
    die "! ERROR: The install directory '$INSTALLROOT' is not writable.\n"
	unless (-w "$INSTALLROOT");
}

if ($opt_help or !@ARGS) {
    print <<"ENDUSAGE";
  Directories:
       temporary: '$SYSTMP/getfont-<PID>'
       install:   '$INSTALLROOT'

ENDUSAGE
check_tmpdir;
check_installroot;
exit 0;
}

check_tmpdir;
check_installroot;

my $tmpdir="$SYSTMP/getfont-$$";
debug_msg "Internal variable tmpdir set to '$tmpdir'";

mkdir "$tmpdir" or die "! ERROR: Can't mkdir '$tmpdir'.";
chdir "$tmpdir" or die "! ERROR: Can't cd '$tmpdir'.";

sub remove_tmpdir {
    chdir "$SYSTMP" or die "! ERROR: Can't cd '$SYSTMP'.\n";
    opendir TMPDIR, "$tmpdir" 
	or print "! ERROR: Can't read directory '$tmpdir'.\n";

    @alltmps=readdir TMPDIR;
    closedir TMPDIR;

    foreach $file (@alltmps) {
	next if $file=~/^\./;
	unlink "$tmpdir/$file";
    }
    rmdir "$tmpdir" or die "! ERROR: Can't remove directory '$tmpdir'.\n";
}

# Signal handlers.

my @common_signals=qw(INT ILL FPE SEGV TERM ABRT);
my @signals_UNIX=qw(QUIT BUS PIPE);
my @signals_Win32=qw(BREAK);


if ($^O=~/^MSWin(32|64)$/i) {
    @signals=(@common_signals, @signals_Win32);
} else {
    @signals=(@common_signals, @signals_UNIX);
}

debug_msg "Supported signals: @signals";

foreach my $signal (@signals) {
    $SIG{"$signal"}=\&remove_tmpdir;
}

# Download the script from tug.org.

my $has_wget=0;

if ($^O=~/^MSWin(32|64)$/i) {
    $has_wget=1; ## shipped with TL.
} elsif (which "wget") {
    $has_wget=1;
}

debug_msg "No wget binary found on your system, trying curl" unless ($has_wget);

if ($has_wget) {
    debug_msg "Running 'wget $getfont_url'";
    system ('wget', "$getfont_url") == 0 
	or die "! Error: Can't execute wget.\n";
} else {
    debug_msg "Running 'curl -O $getfont_url'";
    system ('curl', '-O', "$getfont_url") == 0 
	or die "! Error: Can't execute curl.\n";
}

# Download the fonts.

my @getfont=('perl', "./getfont$TL_version", '--getnonfreefonts');
push @getfont, "--installroot=$INSTALLROOT";
push @getfont, '--lsfonts' if $opt_lsfonts;
push @getfont, '--force' if $opt_force;
push @getfont, '--debug' if $opt_debug;
push @getfont, '--verbose' if $opt_verbose;
push @getfont, '--sys' if $sys;
push @getfont, '--all' if $opt_all;
push @getfont, '--use_curl' unless ($has_wget);
push @getfont, @allpackages;

debug_msg "Running '@getfont'";

system @getfont;

# Evaluate the exit status.  It will be 2 if something had been
# installed and in this case we need to run texhash/updmap.

my $exit_code=$?;
my $exit_status=int($exit_code/256);

if ($sys) {		     
    debug_msg "Info: Execute updmap-sys if exit status is 2";
} else {
    debug_msg "Info: Execute updmap if exit status is 2";
}

debug_msg "Exit status of getfont$TL_version is $exit_status";

if ($exit_status==2) {
    print "\n";
    system 'texhash', "$INSTALLROOT";
    $updmap_command=($sys)? 'updmap-sys':'updmap';
    @updmap=("$updmap_command");
    push @updmap, '--quiet' unless $opt_verbose;
    print "\n";
    message "Updating map files ($updmap_command)...";
    system @updmap;
    
    message $?? '[failed]':'[done]';
}

remove_tmpdir;

__END__
